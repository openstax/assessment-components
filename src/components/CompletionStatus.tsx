import styled from "styled-components";
import { InnerStepCard } from "./Card";
import Button from "./Button";

export interface CompletionStatusProps {
  numberOfQuestions: number;
  numberCompleted: number;
  canContinue: boolean;
  handleClick: () => void;
}

const CompletionStatusCard = styled(InnerStepCard)`
  padding: 88px 72px;
  font-size: 1.8rem;
  line-height: 3rem;
  display: block;

  button {
    min-width: 160px;
    height: 48px;
  }

  p {
    margin: 16px 0 20px 0;
  }
`;

const CompletionHeader = styled.h2`
  font-size: 2.4rem;
  margin: 0;
`;

export const CompletionStatus = ({
  numberOfQuestions, numberCompleted, canContinue, handleClick
}: CompletionStatusProps) => {
  
  const allCompleted = numberOfQuestions === numberCompleted;
  const someCompleted = numberCompleted > 0;
  const buttonText = allCompleted && canContinue ? 'Continue to next step' : (
      someCompleted ? 'Continue' : 'Start'
  );

  return (
    <CompletionStatusCard>
      <CompletionHeader>{allCompleted ? 'You are done.' : (someCompleted ? 'Steps in progress' : 'No questions have been answered.')}</CompletionHeader>
      <p>{allCompleted ? 'Great job answering all the questions.' : (someCompleted ? `You've completed ${numberCompleted} of ${numberOfQuestions} questions.` : 'Begin working on the quiz.')}</p>
      {allCompleted && !canContinue ? null :
        <Button data-test-id={`${buttonText.split(' ')[0].toLowerCase()}-btn`} onClick={()=> handleClick()}>
          {buttonText}
        </Button>
      }
    </CompletionStatusCard>
  )
};
