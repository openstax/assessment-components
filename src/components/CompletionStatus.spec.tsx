import { CompletionStatus, CompletionStatusProps } from './CompletionStatus';
import Button from './Button';
import renderer from 'react-test-renderer';

describe('CompletionStatus', () => {
  let props: CompletionStatusProps;

  beforeEach(() => {
    props = {
      numberOfQuestions: 15,
      numberCompleted: 0,
      handleClick: () => {console.log('click')}
    }
  });

  it('matches snapshot', () => {
    const tree = renderer.create(
      <CompletionStatus {...props} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with some questions completed', () => {
    const tree = renderer.create(
      <CompletionStatus {...props} numberCompleted={5} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with all questions completed', () => {
    const tree = renderer.create(
      <CompletionStatus {...props} numberCompleted={15} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('clicking triggers handler', () => {
    const mockEv = jest.fn();
    const component = renderer.create(
      <CompletionStatus {...props} handleClick={mockEv} />
    );

    renderer.act(() => {
      component.root.findByType(Button).props.onClick();
    });

    expect(mockEv).toHaveBeenCalled();
  })
});
