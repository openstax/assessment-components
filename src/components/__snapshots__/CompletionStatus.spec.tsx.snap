// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompletionStatus matches snapshot 1`] = `
<div
  className="sc-bczRLJ sc-jSMfEi YKyJl kikmgW"
>
  <h2
    className="sc-gKXOVf bebcLz"
  >
    No questions have been answered.
  </h2>
  <p>
    Begin working on the quiz.
  </p>
  <button
    className="sc-eCYdqJ dqiaKp"
    data-test-id="start-btn"
    onClick={[Function]}
  >
    Start
  </button>
</div>
`;

exports[`CompletionStatus matches snapshot with all questions completed 1`] = `
<div
  className="sc-bczRLJ sc-jSMfEi YKyJl kikmgW"
>
  <h2
    className="sc-gKXOVf bebcLz"
  >
    You are done.
  </h2>
  <p>
    Great job answering all the questions.
  </p>
</div>
`;

exports[`CompletionStatus matches snapshot with all questions completed and continue enabled 1`] = `
<div
  className="sc-bczRLJ sc-jSMfEi YKyJl kikmgW"
>
  <h2
    className="sc-gKXOVf bebcLz"
  >
    You are done.
  </h2>
  <p>
    Great job answering all the questions.
  </p>
  <button
    className="sc-eCYdqJ dqiaKp"
    data-test-id="continue-btn"
    onClick={[Function]}
  >
    Continue to next step
  </button>
</div>
`;

exports[`CompletionStatus matches snapshot with some questions completed 1`] = `
<div
  className="sc-bczRLJ sc-jSMfEi YKyJl kikmgW"
>
  <h2
    className="sc-gKXOVf bebcLz"
  >
    Steps in progress
  </h2>
  <p>
    You've completed 5 of 15 questions.
  </p>
  <button
    className="sc-eCYdqJ dqiaKp"
    data-test-id="continue-btn"
    onClick={[Function]}
  >
    Continue
  </button>
</div>
`;
